### ISSUE TITLE
Controlled Experiments in Machine Learning

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#461]
Controlled Experiments in Machine Learning<br>
<br>
https://ift.tt/2Iabnzy<br>
<br>
Jason Brownlee<br>
<br>
<p>Systematic experimentation is a key part of applied machine learning.</p>
<p>Given the complexity of machine learning methods, they resist formal analysis methods. Therefore, we must learn about the behavior of algorithms on our specific problems empirically. We do this using controlled experiments.</p>
<p>In this tutorial, you will discover the important role that controlled experiments play in applied machine learning.</p>
<p>After completing this tutorial, you will know:</p>
<ul>
<li>The need for systematic discovery via controlled experiments.</li>
<li>The need to repeat experiments in order to control for the sources of variance.</li>
<li>Examples of experiments performed in machine learning and the challenge and opportunity they represent.</li>
</ul>
<p>Let&rsquo;s get started.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/06/Controlled-Experiments-in-Machine-Learning.jpg" alt="Controlled Experiments in Machine Learning" width="640" class="wp-image-5637" height="426">
<p class="wp-caption-text">Controlled Experiments in Machine Learning<br>
Photo by <a href="https://www.flickr.com/photos/mikebaird/1723372755/">Mike Baird</a>, some rights reserved.</p>
</div>
<h2>Tutorial Overview</h2>
<p>This tutorial is divided into 3 parts; they are:</p>
<ol>
<li>Systematic Experimentation</li>
<li>Controlling For Variance</li>
<li>Experiments in Machine Learning</li>
</ol>
<center>
<h3>Need help with Statistics for Machine Learning?</h3>
<p>Take my free 7-day email crash course now (with sample code).</p>
<p>Click to sign-up and also get a free PDF Ebook version of the course.</p>
<p><a href="https://machinelearningmastery.lpages.co/leadbox/142f75173f72a2%3A164f8be4f346dc/5750943224168448/" target="_blank">Download Your FREE Mini-Course</a></p>
</center>
<h2>Systematic Experimentation</h2>
<p>In applied machine learning, you must become the scientist and perform systematic experiments.</p>
<p>The answers to questions that you care about, such as what algorithm works best on your data or which input features to use, can only be found through the results of experimental trials.</p>
<p>This is due mainly to the fact that machine learning methods are complex and resist formal methods of analysis.</p>
<blockquote>
<p>[&hellip;] many learning algorithms are too complex for formal analysis, at least at the level of generality assumed by most theoretical treatments. As a result, empirical studies of the behavior of machine learning algorithms must retain a central role.</p>
</blockquote>
<p>&mdash; <a href="https://dl.acm.org/citation.cfm?id=637939">The Experimental Study of Machine Learning</a>, 1991.</p>
<p>In statistics, the choice of a type of experiment is called <a href="https://en.wikipedia.org/wiki/Design_of_experiments">experimental design</a>, and there are many types of experiments to choose from. For example, you may have heard that&nbsp;<a href="https://en.wikipedia.org/wiki/Blinded_experiment">randomized double-blind placebo-controlled experimentation</a> as the gold standard for evaluating the effectiveness of medical treatments.</p>
<p>Applied machine learning is special in that we have complete control over the experiment and we can run as few or as many trials as we wish on our computer. Because of the ease of running experiments, it is important that we are running the right types of experiments.</p>
<blockquote>
<p>In the natural sciences, one can never control all possible variables. [&hellip;] As a science of the artificial, machine learning can usually avoid such complications.</p>
</blockquote>
<p>&mdash; <a href="https://link.springer.com/article/10.1023%2FA%3A1022623814640?LI=true">Machine Learning as an Experimental Science</a>, Editorial, 1998.</p>
<p>The type of experiments we wish to perform are called <a href="https://en.wikipedia.org/wiki/Scientific_control">controlled experiments</a>.</p>
<p>These are experiments where all known independent variables are held constant and modified one at a time in order to determine their impact on the dependent variable. The results are compared to a baseline, or no-treatment, called a &ldquo;<em>control</em>.&rdquo; This could be the result of a baseline method like persistence or the Zero Rule algorithm or the default-configuration for the method.</p>
<blockquote>
<p>As normally defined, an experiment involves systematically varying one or more independent variables and examining their effect on some dependent variables. Thus, a machine learning experiment requires more than a single learning run; it requires a number of runs carried out under different conditions. In each case, one must measure some aspect of the system&rsquo;s behavior for comparison across the different conditions.</p>
</blockquote>
<p>&mdash; <a href="https://link.springer.com/article/10.1023%2FA%3A1022623814640?LI=true">Machine Learning as an Experimental Science</a>, Editorial, 1998.</p>
<h2>Controlling For Variance</h2>
<p>In many ways, experiments with machine learning methods have more in common with <a href="https://en.wikipedia.org/wiki/Computer_experiment">simulation studies</a>, such as those in physics, than with evaluating medical treatments.</p>
<p>As such, the results of a single experiment are probabilistic, subjected to variance.</p>
<p>There are two main types of variance that we seek to understand in our controlled experiments; they are:</p>
<ul>
<li>
<strong>Variance in the data</strong>, such as the data used to train the learning algorithm and the data used to evaluate its skill.</li>
<li>
<strong>Variance in the model</strong>, such as the use of randomness in the learning algorithm, such as random initial weights in neural nets, selection of cut points in bagging, shuffled order of data in stochastic gradient descent, and so on.</li>
</ul>
<p>A result from a single run or trial of a controlled experiment would be misleading given these sources of variance.</p>
<p>The experiment must control for these sources of variance. This is done by repeating the experimental trial multiple times in order to elicit the range of variance so that we can both report the expected result and the variance in the expected result, e.g. mean and confidence interval.</p>
<p>In simulation studies, such as Monte Carlo methods, the repetition of an experiment is called <a href="https://en.wikipedia.org/wiki/Variance_reduction">variance reduction</a>.</p>
<h2>Experiments in Machine Learning</h2>
<p>Experimentation is a key part of applied machine learning.</p>
<p>This is both a challenge to beginners who must learn some rigor and an exciting opportunity for discovery and contribution.</p>
<p>Let&rsquo;s make this concrete with some examples of the types of controlled experiments you may need to perform:</p>
<ul>
<li>
<strong>Choose-Features Experiments</strong>. When determining what data features (input variables) are most relevant to a model, the independent variables may be the input features and the dependent variable might be the estimated skill of the model on unseen data.</li>
<li>
<strong>Tune-Model Experiments</strong>. When tuning a machine learning model, the independent variables may be the hyperparameters of the learning algorithm and the dependent variable might be the estimated skill of the model on unseen data.</li>
<li>
<strong>Compare-Models Experiments</strong>. When comparing the performance of machine learning models, the independent variables may be the learning algorithms themselves with a specific configuration and the dependent variable is the estimated skill of the model on unseen data.</li>
</ul>
<p>What makes the experimental focus of applied machine learning so exciting is two fold:</p>
<ul>
<li>
<strong>Discovery</strong>. You can discover what works best for your specific problem and data. A challenge and an opportunity.</li>
<li>
<strong>Contribution</strong>. You can make broader discoveries in the field, without any specialized knowledge other than rigorous and systematic experimentation.</li>
</ul>
<p>Using off-the-shelf tools and careful experimental methods, you can make discoveries and contributions.</p>
<blockquote>
<p>In summary machine learning occupies a fortunate position that makes systematic experimentation easy and profitable. [&hellip;] Although experimental studies are not the only path to understanding, we feel they constitute one of machine learning s brightest hopes for rapid scientific progress, and we encourage other researchers to join in our fields evolution toward an experimental science.</p>
</blockquote>
<p>&mdash; <a href="https://dl.acm.org/citation.cfm?id=637939">The Experimental Study of Machine Learning</a>, 1991.</p>
<h2>Further Reading</h2>
<p>This section provides more resources on the topic if you are looking to go deeper.</p>
<h3>Books</h3>
<ul>
<li>
<a href="https://amzn.to/2Ge4yAP">The Design and Analysis of Computer Experiments</a>, 2003.</li>
<li>
<a href="https://amzn.to/2GdUPut">Empirical Methods for Artificial Intelligence</a>, 1995.</li>
</ul>
<h3>Papers</h3>
<ul>
<li>
<a href="https://link.springer.com/article/10.1023%2FA%3A1022623814640?LI=true">Machine Learning as an Experimental Science</a>, Editorial, 1998.</li>
<li>
<a href="https://dl.acm.org/citation.cfm?id=637939">The Experimental Study of Machine Learning</a>, 1991.</li>
<li>
<a href="http://www.aaai.org/Papers/Workshops/2006/WS-06-06/WS06-06-002.pdf">Machine Learning as an Experimental Science (Revisited)</a>, 2006.</li>
</ul>
<h3>Articles</h3>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Scientific_control">Scientific control on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Design_of_experiments">Design of experiments on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Blinded_experiment">Blinded experiment on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Controlling_for_a_variable">Controlling for a variable on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Computer_experiment">Computer experiment on Wikipedia</a></li>
<li><a href="https://en.wikipedia.org/wiki/Variance_reduction">Variance Reduction on Wikipedia</a></li>
</ul>
<h3>Summary</h3>
<p>In this tutorial, you discovered the important role that controlled experiments play in applied machine learning.</p>
<p>Specifically, you learned:</p>
<ul>
<li>The need for systematic discovery via controlled experiments.</li>
<li>The need to repeat experiments in order to control for the sources of variance.</li>
<li>Examples of experiments performed in machine learning and the challenge and opportunity they represent.</li>
</ul>
<p>Do you have any questions?<br>
Ask your questions in the comments below and I will do my best to answer.</p>
<p>The post <a rel="nofollow" href="https://machinelearningmastery.com/controlled-experiments-in-machine-learning/">Controlled Experiments in Machine Learning</a> appeared first on <a rel="nofollow" href="https://machinelearningmastery.com">Machine Learning Mastery</a>.</p>
<br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning Mastery https://ift.tt/2if3vo8<br>
<br>
June 22, 2018 at 03:10AM
